### PR TITLE
2027 adjustable ROI

### LINKED ISSUE #2027: Adjustable ROI in Spectrum Viewer
### Desired Behaviour

In the Spectrum Viewer, it would be useful to be able to see the details of the ROI currently on the screen, i.e. its placement/dimensions. It may also be useful to plot other details about the ROI, i.e. number of pixels in ROI or any other "on the fly data/calculation".

It would also be an advantage to be able to use this interface to be able to make small pixel-by-pixel adjustments if the user wants to put the ROI in a very specific place without zooming in and dragging with the mouse.

### Current Behaviour

There is no information displayed in the GUI about the ROI.

### PR DESCRIPTION
**Pull request by MikeSullivan7**
### Issue

Closes #2027.

### Description

A table has been added under the ROI tabs in the Spectrum Viewer which displays the properties of the currently selected ROI (selected by clicking on the ROI Table). Currently the properties displayed are: Left, Top, Right, Bottom, Width, Height, Area.
The currently selected ROI can be adjusted by using spin boxes in the table allowing for more precise positioning.
This table also functions for the RITS ROI when that tab is activated.

More properties and details about the ROIs can be added as needed/required.

Unit Tests will be written in a future commit.

image

### Testing 

make check

### Acceptance Criteria 

Check that the numbers in the table are correct and update accordingly (i.e. when the ROI is changed).
Check that changing the numbers in the Table update the ROI properly.
The rest of the Spectrum viewer should function the same as before.

### CODE REVIEW

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 1]
```diff
@@ -92,7 +97,47 @@ def __init__(self, main_window: 'MainWindowView'):
         self.tableView.setSelectionMode(QAbstractItemView.SingleSelection)
         self.tableView.setAlternatingRowColors(True)
 
+        # Roi Prop table
+        self.roiPropertiesGroupBoxTitle: str = "Roi Properties: "
+        self.roi_table_properties = ["Top", "Bottom", "Left", "Right"]
+        self.roi_table_properties_secondary = ["Width", "Height", "Area"]
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:103]: If @samtygier-stfc and maybe IMAT agree, i'd consider removing area if not needed to free up some space in the GUI.

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 2]
```diff
@@ -334,3 +384,40 @@ def set_binning_visibility(self) -> None:
...
+        spin_boxes = self.roiPropertiesSpinBoxes
+        for key, value in spin_boxes.items():
+            if key == "Left":
+                value.setMaximum(spin_boxes["Right"].value() - 1)
+            elif key == "Right":
+                value.setMinimum(spin_boxes["Left"].value() + 1)
+            elif key == "Top":
+                value.setMaximum(spin_boxes["Bottom"].value() - 1)
+            elif key == "Bottom":
+                value.setMinimum(spin_boxes["Top"].value() + 1)
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:423]: Maybe a silly question, but it this needed? I tried commenting out entirely including references and all seemed to perform fine.
If it is needed, you may be able to condense to something like:
```python
        self.roiPropertiesSpinBoxes["Left"].setMaximum(self.roiPropertiesSpinBoxes["Right"].value() - 1)
        self.roiPropertiesSpinBoxes["Right"].setMinimum(self.roiPropertiesSpinBoxes["Left"].value() + 1)
        self.roiPropertiesSpinBoxes["Top"].setMaximum(self.roiPropertiesSpinBoxes["Bottom"].value() - 1)
        self.roiPropertiesSpinBoxes["Bottom"].setMinimum(self.roiPropertiesSpinBoxes["Top"].value() + 1)
```

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 3]
```diff
@@ -334,3 +384,40 @@ def set_binning_visibility(self) -> None:
...
+        row = 0
+        for pos in current_roi.__iter__():
+            with QSignalBlocker(self.roiPropertiesSpinBoxes[roi_iter_order[row]]):
+                self.roiPropertiesSpinBoxes[roi_iter_order[row]].setValue(pos)
+            row = row + 1
+        self.set_roi_spinbox_ranges()
+        self.presenter.redraw_spectrum(self.current_roi)
+        roi_width = abs(self.roiPropertiesSpinBoxes["Right"].value() - self.roiPropertiesSpinBoxes["Left"].value())
+        self.roiPropertiesLabels["Width"].setText(str(roi_width))
+        roi_height = abs(self.roiPropertiesSpinBoxes["Top"].value() - self.roiPropertiesSpinBoxes["Bottom"].value())
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:401]: `roi_width` and `roi_height` can just be accessed using `current_roi.width` and `current_roi.height` respectively to save you calculating it.

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 4]
```diff
@@ -334,3 +384,40 @@ def set_binning_visibility(self) -> None:
...
+        self.roiPropertiesLabels["Height"].setText(str(roi_height))
+        roi_area = roi_width * roi_height
+        self.roiPropertiesLabels["Area"].setText(str(roi_area))
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:404]: Can drop if area is not needed pending @samtygier-stfc and IMAT

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 5]
```diff
@@ -92,7 +97,47 @@ def __init__(self, main_window: 'MainWindowView'):
...
+
+        self.roiPropertiesSpinBoxes = {}
+        self.roiPropertiesLabels = {}
+        row_ind = 0
+        for prop in self.roi_table_properties_all:
+            table_widget = QTableWidgetItem(prop)
+            table_widget.setFlags(Qt.ItemIsSelectable)
+            self.roiPropertiesTableWidget.setItem(row_ind, 0, table_widget)
+            if prop in self.roi_table_properties:
+                spin_box = QSpinBox()
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:116]: Can we decrease the width of the spin boxes? maybe by as much as half to free up the GUI a little bit?

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 6]
```diff
@@ -92,7 +97,47 @@ def __init__(self, main_window: 'MainWindowView'):
...
+                if prop == "Top" or prop == "Bottom":
+                    spin_box.setMaximum(self.spectrum.image.image_data.shape[0])
+                if prop == "Left" or prop == "Right":
+                    spin_box.setMaximum(self.spectrum.image.image_data.shape[1])
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:120]: NITPICK: to use up space more effectively, maybe we could place top and bottom on one row, left and right on one row and width and height on the same row and if Area is needed, Area on one row at the bottom?

File: mantidimaging/gui/ui/spectrum_viewer.ui [hunk 7]
```diff
@@ -6,8 +6,8 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>921</width>
-    <height>739</height>
+    <width>913</width>
+    <height>710</height>
```
Comment (JackEAllen) [inline mantidimaging/gui/ui/spectrum_viewer.ui:10]: If still required after decreasing the width of the spinbox's, can we increase the default size of the window a little more to cover lower resolution screens? I get scroll bars with the initial window size with resolutions 1792 x 1120 x 1080, 1200 and 1080 x 1920
image

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 3 above)
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:401]: NITPICK: If not too hard and straightforward to do, could we place the width and height at the bottom of the image view and above the spectrum plot inline with the "_x_", "_y_" and "_value_", values (see red text in below image). Sam can properly help you with this if you struggle as I believe he did this in the main window. 

Screenshot 2024-02-12 at 08 38 58

**Review by JackEAllen — CHANGES_REQUESTED** [review]
If you're happy to implement it, when an ROI is clicked in the image view, I think we should select that ROI within the ROI table as this seems like it should already be the case intuitively.

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 2 above)
Comment (MikeSullivan7) [reply mantidimaging/gui/windows/spectrum_viewer/view.py:423]: This is mainly to prevent the user from selecting for example Top > Bottom via the spinboxes as it doesnt make much sense. Doing so doesn't break anything but it does give errors in the terminal. 
image
image

I think its worth keeping but I can condense it down.

**Comment by JackEAllen** [comment]
Looks like Applitools tests need to be updated: https://eyes.applitools.com/app/test-results/00000251694396968241/00000251694396790667?accountId=-yE-mpxluECTdfURli4jgA~~

**Comment by JackEAllen** [comment]
Can you increase the minimum height and width of the container to remove the vertical scroll on opening spectrum viewer and prevent cut-off the "Width" label
image

File: mantidimaging/gui/windows/spectrum_viewer/spectrum_widget.py [hunk 8]
```diff
@@ -183,6 +187,14 @@ def add_roi(self, roi: SensibleROI, name: str) -> None:
         self.image.vb.addItem(self.roi_dict[name])
         self.roi_dict[name].hoverPen = mkPen(self.roi_dict[name].colour, width=3)
 
+    def adjust_roi(self, new_roi: SensibleROI, roi_name: str):
+        """
+        Adjust the existing ROI with the given name.
+        @param spec_roi: The new SpectrumROI to replace the existing SpectrumROI
```
[the remark this answers is not in the document]
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/spectrum_widget.py:213]: This still needs updating to match param name

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 9]
```diff
@@ -371,3 +435,30 @@ def set_binning_visibility(self) -> None:
...
+        self.presenter.redraw_spectrum(self.current_roi)
+        self.roiPropertiesLabels["Width"].setText(str(current_roi.width))
+        self.roiPropertiesLabels["Height"].setText(str(current_roi.height))
+
+    def adjust_roi(self) -> None:
+        roi_iter_order = ["Left", "Top", "Right", "Bottom"]
+        new_points = [self.roiPropertiesSpinBoxes[prop].value() for prop in roi_iter_order]
+        new_roi = SensibleROI().from_list(new_points)
+        self.presenter.model.set_roi(self.current_roi, new_roi)
+        self.spectrum.adjust_roi(new_roi, self.current_roi)
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:458]: Should this be: 
```suggestion
        self.adjust_roi(new_roi, self.current_roi)
```

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 3 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/view.py]: Did you find out how to do this - maybe from Sam?

**Review by JackEAllen — CHANGES_REQUESTED** [review]
Thanks for condensing. just a few little bits left now

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 3 above)
Comment (samtygier-stfc) [reply mantidimaging/gui/windows/spectrum_viewer/view.py]: Having been looking at the message setting code recently, I'd see leave it in the side panel for now.

At some we should add a way to make it easier for a window to customise what is shown in the image view message.

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 9 above)
Comment (MikeSullivan7) [reply mantidimaging/gui/windows/spectrum_viewer/view.py:458]: No, this is in self.adjust_roi() so that would be circular. It should be `self.spectrum_widget.adjust_roi(new_roi, self.current_roi)` to adjust the spectrum roi in the spectrum widget

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 3 above)
Comment (MikeSullivan7) [reply mantidimaging/gui/windows/spectrum_viewer/view.py]: Yes I think that its something that can be looked at after release as a small change (or potentially big change if it leads to a rabbit hole)

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 10]
```diff
@@ -371,3 +434,28 @@ def set_binning_visibility(self) -> None:
         self.bin_size_spinBox.setHidden(hide_binning)
         self.bin_step_label.setHidden(hide_binning)
         self.bin_step_spinBox.setHidden(hide_binning)
+
+    def set_roi_properties(self) -> None:
+        current_roi = self.presenter.model.get_roi(self.current_roi)
+        self.roiPropertiesGroupBox.setTitle(f"Roi Properties: {self.current_roi}")
+        roi_iter_order = ["Left", "Top", "Right", "Bottom"]
+        for row, pos in enumerate(current_roi):
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:442]: If I delete all ROI's within ROI mode, We should clear the values from the spin boxes:
image

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 11]
```diff
@@ -98,7 +107,55 @@ def __init__(self, main_window: 'MainWindowView'):
...
+            item.setFlags(Qt.ItemIsSelectable)
+            self.roiPropertiesTableTextDict[text] = item
+
+        self.roiPropertiesTableWidget.setItem(0, 0, self.roiPropertiesTableTextDict["x1, x2"])
+        self.roiPropertiesTableWidget.setCellWidget(0, 1, self.roiPropertiesSpinBoxes["Left"])
+        self.roiPropertiesTableWidget.setCellWidget(0, 2, self.roiPropertiesSpinBoxes["Right"])
+        self.roiPropertiesTableWidget.setItem(1, 0, self.roiPropertiesTableTextDict["y1, y2"])
+        self.roiPropertiesTableWidget.setCellWidget(1, 1, self.roiPropertiesSpinBoxes["Top"])
+        self.roiPropertiesTableWidget.setCellWidget(1, 2, self.roiPropertiesSpinBoxes["Bottom"])
+        self.roiPropertiesTableWidget.setItem(2, 0, self.roiPropertiesTableTextDict["Width, Height"])
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:150]: This currently does not fit the default width. Maybe reword to "Size"
image

**Review by JackEAllen — CHANGES_REQUESTED** [review]
Looks good! just a few more tiny things and then this is good to merge

**Comment by JackEAllen** [comment]
I'd recommend rebasing to update you're branch with main to double check the bug isn't implemented by your work as the bug doesn't exist on main and I don't believe it is the proposed changes on this branch causing the bug, but if you remove all ROI's and swap to the image mode and the return back to the ROI mode, can your replicate experiencing a KeyError exception where the default ROI is not re-added after swapping views.? 
image

**Comment by MikeSullivan7** [comment]
> I'd recommend rebasing to update you're branch with main to double check the bug isn't implemented by your work as the bug doesn't exist on main and I don't believe it is the proposed changes on this branch causing the bug, but if you remove all ROI's and swap to the image mode and the return back to the ROI mode, can your replicate experiencing a KeyError exception where the default ROI is not re-added after swapping views.? image

This bug should now be fixed now, I changed the table behaviour slightly when the tabs are switched back and forth even with all ROIs removed. I do not see this error when testing so can you please check whether you still see this bug?

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 3 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/view.py]: See: #2052

**Review by JackEAllen — CHANGES_REQUESTED** [review]
Thanks for adding all the changes in! Looks really good! 
Just two small things now and we're there! 👍 

If I delete all ROIs from the ROI view, then navigate to the Image view, the spin boxes need to be re-enabled for that view then disabled again on returning to Image view if no ROI's exist (See screenshot)

image


If i remove all ROI's from ROI view, navigate to Image view, swap samples, then return to ROI view, the state changes are not synced up causing the KeyError (see screenshot and logs)

image

Logs:
```bash
2024-02-16 13:59:44,642 [mantidimaging.core.utility.progress_reporting.progress:L241] INFO: Elapsed time: 11 sec.
2024-02-16 14:00:00,263 [perf.mantidimaging.gui.mvp_base.view:L59] INFO: SpectrumViewerWindowView shown in 0.8960616621188819
2024-02-16 14:00:10,615 [py.warnings:L109] WARNING: /home/vco13151/mambaforge/envs/mantidimaging-dev/lib/python3.10/site-packages/pyqtgraph/graphicsItems/PlotItem/PlotItem.py:509: UserWarning: Item already added to PlotItem, ignoring.

2024-02-16 14:00:13,657 [mantidimaging.gui.windows.main.view:L513] ERROR: Traceback (most recent call last):
  File "/home/vco13151/mantidimaging/mantidimaging/gui/windows/spectrum_viewer/presenter.py", line 288, in handle_export_tab_change
    self.view.on_visibility_change()
  File "/home/vco13151/mantidimaging/mantidimaging/gui/windows/spectrum_viewer/view.py", line 221, in on_visibility_change
    self.set_roi_properties()
  File "/home/vco13151/mantidimaging/mantidimaging/gui/windows/spectrum_viewer/view.py", line 450, in set_roi_properties
    current_roi = self.presenter.model.get_roi(self.current_roi)
  File "/home/vco13151/mantidimaging/mantidimaging/gui/windows/spectrum_viewer/model.py", line 122, in get_roi
    raise KeyError(f"ROI {roi_name} does not exist in roi_ranges {self._roi_ranges.keys()}")
KeyError: "ROI None does not exist in roi_ranges dict_keys(['all', 'roi', 'rits_roi'])"

2024-02-16 14:00:13,657 [mantidimaging.gui.mvp_base.view:L40] ERROR: show_error_dialog(): Uncaught exception KeyError: "ROI None does not exist in roi_ranges dict_keys(['all', 'roi', 'rits_roi'])"


```

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 12]
```diff
@@ -371,3 +445,44 @@ def set_binning_visibility(self) -> None:
         self.bin_size_spinBox.setHidden(hide_binning)
         self.bin_step_label.setHidden(hide_binning)
         self.bin_step_spinBox.setHidden(hide_binning)
+
+    def set_roi_properties(self) -> None:
+        if not any([spinbox.setEnabled(False) for spinbox in self.roiPropertiesSpinBoxes.values()]):
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:450]: This causes ruff checks to fail under code: `C419 Unnecessary list comprehension`, I'd suggest changing to this to pass Ruff static analysis
```suggestion
    for spinbox in self.roiPropertiesSpinBoxes.values():
        spinbox.setEnabled(True)
```

**Review by JackEAllen — CHANGES_REQUESTED** [review]

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 13]
```diff
@@ -371,3 +445,46 @@ def set_binning_visibility(self) -> None:
...
+
+    def set_roi_spinbox_ranges(self):
+        self.roiPropertiesSpinBoxes["Left"].setMaximum(self.roiPropertiesSpinBoxes["Right"].value() - 1)
+        self.roiPropertiesSpinBoxes["Right"].setMinimum(self.roiPropertiesSpinBoxes["Left"].value() + 1)
+        self.roiPropertiesSpinBoxes["Top"].setMaximum(self.roiPropertiesSpinBoxes["Bottom"].value() - 1)
+        self.roiPropertiesSpinBoxes["Bottom"].setMinimum(self.roiPropertiesSpinBoxes["Top"].value() + 1)
+
+    def disable_roi_properties(self):
+        self.roiPropertiesGroupBox.setTitle("Roi Properties: None selected")
+        self.last_clicked_roi = "roi"
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:480]: When changing samples within the Image mode, the new ROI generated falls out of sync with the ROI that the spin boxes are watching such that when I resize the ROI within the image view, the spin box values do not update.

**Review by JackEAllen — CHANGES_REQUESTED** [review]
Just this to fix now and good to merge
